File: aafixer.lua
require 'lfs'

-- Parse arguments
inputpath = arg[2]
outputpath = arg[4]

-- Ensure directory paths end in slashes
if string.sub(inputpath, -1) ~= '/' then
   inputpath = inputpath .. '/'
end

if string.sub(outputpath, -1) ~= '/' then
   outputpath = outputpath .. '/'
end

notbigenough = {}
notsquare = {}

-- Main part of script
for artist in lfs.dir(inputpath) do
   if artist ~= "." and artist ~= ".." then
      for album in lfs.dir(inputpath .. artist) do
         if album ~= "." and album ~= ".." then
            for aart in lfs.dir(inputpath .. artist .. '/' .. album) do
               if aart:match("AlbumArt%.jpg$") then
                  ffpath = inputpath .. artist .. '/' .. album .. '/' .. aart
                  ch = io.popen('magick identify "' .. ffpath .. '"')
                  result = ch:read('*a')
                  dims = string.match(result, "%d+x%d+")
                  aadims = {}
                  for m in dims:gmatch("%d+") do
                     table.insert(aadims, m)
                  end
                  if aadims[1] != aadims[2] then
                     table.insert(notsquare, ffpath)
                  end
                  if (tonumber(aadims[1]) < 1000) or (tonumber(aadims[2]) < 1000) then
                     table.insert(notbigenough, ffpath)
                  end
               end
            end
         end
      end
   end
end
fbe = io.open(outputpath .. "notbig.txt", "w")
fns = io.open(outputpath .. "notsquare.txt", "w")
for i = 1, #notbigenough do
   fbe:write(notbigenough[i], "\n")
end
for i = 1, #notsquare do
   fns:write(notsquare[i], "\n")
end
fns:close()
fbe:close()

File: aafixer.nim
#!/usr/bin/env nim
import os, osproc, parseopt, re, strutils

# Parse arguments
var options: tuple[inputPath: string, outputPath: string]
for kind, key, val in getopt():
  case kind
  of cmdArgument:
    continue
  of cmdLongOption, cmdShortOption:
    case key
    of "input", "i": options.inputPath = val
    of "output", "o": options.outputPath = val
  of cmdEnd: assert(false)

# Ensure directory paths end in slashes
if options.inputPath[^1] != '/':
  options.inputPath = options.inputPath & "/"
if options.outputPath[^1] != '/':
  options.outputPath = options.outputPath & "/"

# Variable declaration
var notSquare = newSeq[string](0)
var notBigEnough = newSeq[string](0)

# Main part of program
for file in walkDirRec(options.inputPath):
  if file.contains(re"AlbumArt") == true:
    var result = execProcess("magick identify \"$1\"" % [file])
    var aadims = result.findAll(re"\d+x\d+")[0].split(re"x")
    if aadims[0] == aadims[1]:
      notSquare.add(file)
    if parseInt(aadims[0]) > 1000 or parseInt(aadims[1]) > 1000:
      notBigEnough.add(file)

# Write results of search to files
var fbe = open(options.outputPath & "notbig.txt", fmWrite)
var fns = open(options.outputPath & "notsquare.txt", fmWrite)
for fi in notBigEnough:
  fbe.writeLine(fi)
for fl in notSquare:
  fns.writeLine(fl)
fns.close()
fbe.close()
